src/util/settings.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/settings.h>
+
+#include <univalue.h>
+
+namespace util {
+namespace {
+
+enum class Source {
+   FORCED,
+   COMMAND_LINE,
+   CONFIG_FILE_NETWORK_SECTION,
+   CONFIG_FILE_DEFAULT_SECTION
+};
+
+//! Merge settings from multiple sources in precedence order:
+//! Forced config > command line > config file network-specific section > config file default section
+//!
+//! This function is provided with a callback function fn that contains
+//! specific logic for how to merge the sources.
+template <typename Fn>
+static void MergeSettings(const Settings& settings, const std::string& section, const std::string& name, Fn&& fn)
+{
+    // Merge in the forced settings
+    if (auto* value = FindKey(settings.forced_settings, name)) {
+        fn(SettingsSpan(*value), Source::FORCED);
+    }
+    // Merge in the command-line options
+    if (auto* values = FindKey(settings.command_line_options, name)) {
+        fn(SettingsSpan(*values), Source::COMMAND_LINE);
+    }
+    // Merge in the network-specific section of the config file
+    if (!section.empty()) {
+        if (auto* map = FindKey(settings.ro_config, section)) {
+            if (auto* values = FindKey(*map, name)) {
+                fn(SettingsSpan(*values), Source::CONFIG_FILE_NETWORK_SECTION);
+            }
+        }
+    }
+    // Merge in the default section of the config file
+    if (auto* map = FindKey(settings.ro_config, "")) {
+        if (auto* values = FindKey(*map, name)) {
+            fn(SettingsSpan(*values), Source::CONFIG_FILE_DEFAULT_SECTION);
+        }
+    }
+}
+} // namespace
+
+SettingsValue GetSetting(const Settings& settings,
+    const std::string& section,
+    const std::string& name,
+    bool ignore_default_section_config,
+    bool get_chain_name)
+{
+    SettingsValue result;
+    MergeSettings(settings, section, name, [&](SettingsSpan span, Source source) {
+        // Weird behavior preserved for backwards compatibility: Apply negated
+        // setting even if non-negated setting would be ignored. A negated
+        // value in the default section is applied to network specific options,
+        // even though normal non-negated values there would be ignored.
+        const bool never_ignore_negated_setting = span.last_negated();
+
+        // Weird behavior preserved for backwards compatibility: Take first
+        // assigned value instead of last. In general, later settings take
+        // precedence over early settings, but for backwards compatibility in
+        // the config file the precedence is reversed for all settings except
+        // chain name settings.
+        const bool reverse_precedence = (source == Source::CONFIG_FILE_NETWORK_SECTION || source == Source::CONFIG_FILE_DEFAULT_SECTION) && !get_chain_name;
```

Comment:
These long lines are killin' me but I'll bite my tongue in the interest of a merge.

Reply:
Will fix with the other followups, but if it in case its useful https://greasyfork.org/en/scripts/18789-github-toggle-code-wrap seems to work well. (Found through https://github.com/StylishThemes/GitHub-code-wrap)